### ISSUE TITLE
Bold text get duplicated from .getText()

### ISSUE THREAD

**Issue by RicottaZhang** [rk700/PyMuPDF#218]

**Comment by JorjMcKie** [comment]
That is probably not true - please send me the PDF example.

**Comment by JorjMcKie** [comment]
look at this example I just made:
[t.pdf](https://github.com/rk700/PyMuPDF/files/2556881/t.pdf)

It produces the expected result:

```
>>> import fitz
>>> doc=fitz.open("t.pdf")
>>> print(doc[0].getText())
ABC:
abc

>>> 
```

**Comment by JorjMcKie** [comment]
So the problem is your PDF

**Comment by RicottaZhang** [comment]
[rule_4936.pdf](https://github.com/rk700/PyMuPDF/files/2560189/rule_4936.pdf)
This is the pdf bold text duplicated.
I used the latest version.

**Comment by RicottaZhang** [comment]
I used the same PDF on a online pdf to text converter, the result with no duplicate text.

[rule_4936.txt](https://github.com/rk700/PyMuPDF/files/2560635/rule_4936.txt)

**Comment by RicottaZhang** [comment]
I append the pymupdf getText output here
[rule_4936.txt](https://github.com/rk700/PyMuPDF/files/2560684/rule_4936.txt)

**Comment by JorjMcKie** [comment]
Sorry for the delay - just recovered from an electricity outage of 4.5 hours - the third one in 4 days ...

Well, congratulations: **you found a bug!**
But not the bug yu thought. The duplicated text is **not bold** - it has a "text rendering mode" set to 2. This is a rarely used command (`2 Tr` is the coding in a page's `/Contents` command stream), and its definition is "Fill, then stroke text".
The text itself is indeed there only once, but for some reason, my underlying C-library MuPDF is returning it twice with **all text extraction methods**, page.getText(), `mutool draw / convert`, etc.
For PyMuPDF, there is no way to suppress this - for the time being you have to live with it.

I will submit a bug report to Artifex / MuPDF. All we can do at the moment is wait for their reaction. A bug fix can anyway only be expected with the next MuPDF version (like 1.15.0) around April of next year.

What you could do at the moment is suppressing multiple output of text which has a duplicate rectangle ("bbox") according to the following scheme:

* extract the text via `extract = page.getText("dict")`
* iterate through the text blocks of the returned dictionary: these are items of the list `extract["blocks"]` which have a `"type"` of 0.
* each of these text blocks has a list of lines, each having their `"bbox"`.
* Omit lines with duplicate bboxes.

Sorry - that's all I can say at the moment.

**Comment by RicottaZhang** [comment]
Nice to have this workaround!
You save me from ugly regex match.
Thank you!

**Comment by JorjMcKie** [comment]
Another workaround option, maybe less programming effort, but also lower level down into PDF details:

Your problem comes from this _bold simulation_ (and that's all what it is, actually) using the operator `2 Tr`. You could just go and replace any such operator by `0 Tr`.

I successfully tested your PDF with this script. For each page, it first reads and manipulates the commands in the /Contents object as described. The main trick here is
* make sure we get rid of any white spaces by replacing them with exactly 1 space
* write the modified commands back into the PDF (no worries, it is only temporary!)
* **important:** now read the page a second time and extract the text

```python
import fitz
doc = fitz.open("rule_4936.pdf")
for i in range(len(doc)):
    xref = doc[i]._getContents()[0]         # assumes: there is only one!
    cont = doc._getXrefStream(xref)         # get command source
    cont_t = cont.split()                   # remove white spaces from command source
    cont = b" ".join(cont_t)                # reformatted content stream
    cont = cont.replace(b"2 Tr", b"0 Tr")   # remove bold simulation
    doc._updateStream(xref, cont)           # replace original /Contents
    page = doc[i]                           # now re-read the page
    text = page.getText()                   # now extract the text
    out = open("page-%i.txt" % i, "wb")     # and write it out somewhere
    out.write(text.encode())
    out.close()
```

**Comment by JorjMcKie** [comment]
By the books, a page could have more than one /Contents object. If you are concerned that this might be the case, you need to loop through the list of xref numbers contained in `_getContents()` and do this manipulation for every single /Contents object.

**Comment by RicottaZhang** [comment]
This works for me, thanks a lot~
```
    doc = fitz.open(pdf)
    for i in range(len(doc)):
        contents = doc[i]._getContents()
        for j in range(len(contents)):
            xref = contents[j]
            cont = doc._getXrefStream(xref)         # get command source
            cont_t = cont.split()                   # remove white spaces from command source
            cont = b" ".join(cont_t)                # reformatted content stream
            cont = cont.replace(b"2 Tr", b"0 Tr")   # remove bold simulation
            doc._updateStream(xref, cont)           # replace original /Contents
    return ''.join(list(map(lambda x: x.getText(), doc)))

```

**Comment by JorjMcKie** [comment]
**Update:**
Artifex / MuPDF has just confirmed this bug as described above, and entered a fix, which will presumably be included in the next MuPDF version - as anticipated.

**Comment by JorjMcKie** [comment]
BTW - I never thought of using the `map` function in the way you have demonstrated - very clever!

**Comment by RicottaZhang** [comment]
Hahahah, actually a bad demonstrate of my big boss :)
I prefer list comprehension
`''.join([ item.getText() for item in doc])`

**Comment by gmixo** [comment]
It seems this issue happened again, but workaround with replace 2 Tr -> 0 Tr don't work, since there is no more Tr strings in xref sequences of page stream. Any help would be much appreciated.

**Comment by JorjMcKie** [comment]
@gmixo - can you share the PDF? In case of confidentiality concerns your are welcome to use my e-mail.
Please no multi-gigabyte beasts, however.
I am sure, MuPDF also would be interested.

Maybe your PDF page also contains text in Form XObjects - which will of course also be extracted. Those objects have their own `/Contents` objects and might be using the `Tr` operator, too. Just scanning the page contents would not detect this.

As an immediate solution and depending on which variant of `get_text` you use, you can check, whether consecutive text spans have the same text and the same bbox, inwhich case you would omit one of them.

```python
old_text=""
old_bbox=None
blocks = page.get_text("dict, flags=...)["blocks"]  # propose to exclue images via flags= ...
for block in blocks:
    for line in b["lines"]:
        for span in line["spans"]:
            bbox = span["bbox"]
            text = span["text"]
            if text == old_text and bbox == old_bbox:
                continue
            old_bbox = bbox
            old_text = text
            ...
...
```

You can of course also extend the workaround: determine the xrefs of any Form XObjects via `page.get_xobjects` and then hunt and change `2 Tr` -> `0 Tr` occurrences in those streams.

**Comment by gmixo** [comment]
Thank you for such a fast response!
I try decode source of page and there is an example of double text
'''
b' /F7 12 Tf 1 0 0 -1 0 0 Tm 339 -4366 Td <002D> Tj 1 0 0 -1 0 0 Tm /Span << /ActualText <> >> BDC 338.500000 -4366 Td <002D> Tj EMC '
'''
there is a string between BT ET markers that means there is a text representation, which produces a text 22, than I expect receive just 2. I guess <002D> - is code of char '2' here
If this info os not enough, I can send file on your email.

**Comment by JorjMcKie** [comment]
I see. This situation does not fall in any of the categories discussed or resolved so far.
I am afraid there cannot be done much about it. The two characters "2" are not under control of a `Tr` command, and they also do not occupy the same location: the y-coordinates are the same, but the x-coordinates are different.
There is no reliable way to disambiguate the situation: after all it **could** also be an intention to actually print the character two times.
The only way out for you is going one step deeper and use `get_tect("rawdict")` so you will see each single character with its bbox. If then consecutive characters are equal and **_almost_** have the same bbox (by some heuristics), discard the second one.
You could for example check the intersection `bbox1 & bbox2` and see if its area is >= x% of each of the single bboxes, and / or if the bottom y-coordinate of both bboxes is (almost) the same, etc.

**Comment by JorjMcKie** [comment]
You also can use a radical approach and remove everything between `BDC` / `EMC` delimiters.

**Comment by gmixo** [comment]
Thanks for answer. Looks like removing everything between BDC / EMC delimiters is a straightforward solution. Could you explain what is mean BDC / EMC markers? If this approach allow not to lost any meaningful information, it''ll be great. Checking for text bboxes intersection is the plan B

**Comment by JorjMcKie** [comment]
> allow not to lost any meaningful information

... true only in your special case. In general, it means this (p 851 of the PDF manual):

![grafik](https://user-images.githubusercontent.com/8290722/119454202-4af5ac80-bd06-11eb-8888-584b8a088a3a.png)

The two `bytes`string b"BDC" / b"EMC" could also wrap text only shown under circumstances controlled by Optional Content logic and then would look like ``/OC /MC1 BDC``. In your case it is ``/Span <</ActualText<>>> BDC``. **ActualText** is an instruction specifying "replacement" text (**_empty_** in cour case!). The whole construct is used here just as a technicque to specify the same text following again to simulate bold.
All that **_effort_** just to avoid the embedding of another font with bold characters ...

So you could do a more precise check: if you encounter `/Span` followed by something containing `ActualText` followed by `BDC`, **_then_** you won't remove import stuff 😒

**Comment by gmixo** [comment]
Actually I already involved logic for check spans with duplicate text, but in this case text so short so it joined to one span.
Thank you for detail explanations.
